tests/concepts/test_craft_torch.py

```diff
@@ -0,0 +1,315 @@
+    nb_samples = 100
+
+    for input_shape in input_shapes:
+        # Generate a fake dataset
+        x, y = generate_torch_data(input_shape, nb_labels, nb_samples)
+        model = generate_torch_model(input_shape, nb_labels)
+
+        # cut the model in two parts (as explained in the paper)
+        # first part is g(.) our 'input_to_latent' model,
+        # second part is h(.) our 'latent_to_logit' model
+        #
+        # Use 2 different set of indexes to check the behavior for
+        # different activation shapes ; (0, 1) produces an activation.shape
+        # of 4 dims, (-1, -1) leads to an activation.shape of 2 dims
+        # for index_layer_g, index_layer_h in [(0, 1), (-1, -1)]:
+
+        for index_layer_g, index_layer_h in [(2, 2), (-1, -1)]:
+
+            g = nn.Sequential(*list(model.children())[:index_layer_g])
+            h = nn.Sequential(*list(model.children())[index_layer_h:])
+
+            # The activations must be positives
+            assert torch.all(g(x) >= 0.0)
+
+            # Initialize Craft
+            number_of_concepts = 10
+            patch_size = 15
+
+            craft = Craft(input_to_latent_model = g,
+                        latent_to_logit_model = h,
+                        number_of_concepts = number_of_concepts,
+                        patch_size = patch_size,
+                        batch_size = 64,
+                        device = 'cpu')
+
+            # Now we can fit the concept using our images
+            # Focus on class id 0
+            class_id = 0
+            images_preprocessed = x[y.argmax(1)==class_id] # select only images of class 'class_id'
+            crops, crops_u, w = craft.fit(images_preprocessed, class_id)
+
+            # Checking shape of crops, crops_u, w
+            assert crops.shape[2] == crops.shape[3] == patch_size # Check patch sizes
+            assert crops.shape[0] == crops_u.shape[0] # Check numbers of patches
+            assert crops_u.shape[1] == w.shape[0]
+
+            # Importance estimation
+            importances = craft.estimate_importance()
+            assert len(importances) == number_of_concepts
+
+            # Checking the results of transform()
+            images_u = craft.transform(images_preprocessed)
+            if len(images_u.shape) == 4:
+                assert images_u.shape == (images_preprocessed.shape[0],
+                                          images_preprocessed.shape[2]-1,
+                                          images_preprocessed.shape[3]-1,
+                                          number_of_concepts)
+            elif len(images_u.shape) == 2:
+                assert images_u.shape == (images_preprocessed.shape[0], number_of_concepts)
+            else:
+                assert False # Should not happen
+
+def test_wrong_layers():
+    """Ensure that Craft complains when the input models are incompatible"""
+
+    input_shapes = [(3, 32, 32)]
+    nb_labels = 3
+
+    for input_shape in input_shapes:
```

Comment:
Like Lucas pointed out above, maybe remove the loop with just one element?

Reply:
done